### ISSUE TITLE
`unstable feature` error needs to specify feature names for better usability.

### ISSUE THREAD

**Issue by nejucomo** [rust-lang/rust#24125]
**Synopsis:** The `unstable feature` compiler error does not specify the explicit feature, so when many are enabled in a single line, the user cannot tell which subset are unstable.

**What I did:** Tried to build [rust-peg](https://github.com/kevinmehall/rust-peg) with ``rustc 1.0.0-beta (9854143cb 2015-04-02) (built 2015-04-02)``

**What I expected:** A specific error message that let's me know which unstable features I must explicitly enable in order to build ``rust-peg``.

**What happened:** I got the following error ([rust-peg ticket #72](https://github.com/kevinmehall/rust-peg/issues/72)):

```
src/peg_syntax_ext.rs:1:1: 1:107 error: unstable feature
src/peg_syntax_ext.rs:1 #![feature(plugin_registrar, quote, box_syntax, core, collections, rustc_private, box_patterns, str_char)]
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
note: this feature may not be used in the beta release channel
error: aborting due to previous error
Could not compile `peg`.
```

Which feature(s) are unstable? The error message needs to be more specific, otherwise I need to try each combination in the superset. :-<

## Meta:

```
$ rustc --version --verbose
rustc 1.0.0-beta (9854143cb 2015-04-02) (built 2015-04-02)
binary: rustc
commit-hash: 9854143cba679834bc4ef932858cd5303f015a0e
commit-date: 2015-04-02
build-date: 2015-04-02
host: x86_64-unknown-linux-gnu
release: 1.0.0-beta
```

**Comment by steveklabnik** [comment]
Ah ha! So, until right before beta, we used to say "use feature flag x to enable this feature", but since that doesn't work in beta, we removed it, meaning we lost this bit of info.

**Comment by huonw** [comment]
All features are unstable in beta, I believe the compiler is trying to say that the `feature` attribute itself is illegal.

**Comment by steveklabnik** [comment]
While that's true, the wording can certainly be better.

**Comment by hgrecco** [comment]
I was going to report the same thing (even with the same library,  `peg`). In my ignorance I tried to find out which "feature" was not available in beta.

**Comment by Jfromchina** [comment]
I encountered the same problem, but my version is Release1.0.0, is it the latest version? which one is stable Release?
$ rustc main.rs 
main.rs:1:1: 1:25 error: unstable feature
main.rs:1 #![feature(collections)]
          ^~~~~~~~~~~~~~~~~~~~~~~~
note: this feature may not be used in the stable release channel
error: aborting due to previous error

$ rustc --version --verbose
rustc 1.0.0 (a59de37e9 2015-05-13) (built 2015-05-14)
binary: rustc
commit-hash: a59de37e99060162a2674e3ff45409ac73595c0e
commit-date: 2015-05-13
build-date: 2015-05-14
host: x86_64-unknown-linux-gnu
release: 1.0.0

**Comment by steveklabnik** [comment]
@Jfromchina that is stable, yes. If it were beta, it would say 1.0.0-beta, or nightly, 1.0.0-nightly.

**Comment by steveklabnik** [comment]
Nominating. While it's sort of a papercut, it's also pretty poor UX.

**Comment by nikomatsakis** [comment]
cc @brson

**Comment by nikomatsakis** [comment]
triage: P-medium

**Comment by brson** [comment]
This is an easy matter of modifying the `emit_feature_err` function.

**Comment by brson** [comment]
Oh, it's not so easy actually because the test cases are going to be very unhappy with the different messages on different channels.

**Comment by brson** [comment]
I've started a patch.

**Comment by brson** [comment]
Oh, this originates from the 'unstable feature' lint, so this is just the standard lint error message. Harder to fix.

**Comment by brson** [comment]
Probably the unstable_feature lint should be converted to a no-op and this check folded into feature_gate.rs.
